collect_app/src/main/java/org/odk/collect/android/preferences/source/SettingsProvider.kt

```diff
@@ -13,7 +15,7 @@ class SettingsProvider(private val context: Context) {
     }
 
     @JvmOverloads
-    fun getGeneralSettings(projectId: String = ""): Settings {
+    fun getGeneralSettings(projectId: String = NOT_SPECIFIED_UUID): Settings {
```

Comment:
Generally for optional params like this I'd use `null` rather than a constant. Is there any reason not to do that here?

Reply:
I would need to make uuid in Project class nullable what I don't like because if I read projects from the database I shouldn't have null values since those ids are generated if not exist. So I would need to add some `!!` or `?.` in the code to handle it.
This also depends on what we decide about what you said below (whether to always use prefixes or not)

Reply:
Right I see. I think that's a signal that we shouldn't be using the same object both what a project looks like before it is "saved" and how it is stored. In my head there are two ways we could go that would avoid a constant for "not specified":

1. Make project an `abstract` with two inner subclasses: `New` and `Existing`. The former does not have an id field but the latter does. `ProjectsRepository#save` could take a `Project` (seeing as our repositories usually do insert and update from save) but return a `Projects.Existing`. `Project.Existing` objects would be stored using GSON.
2. Create a private class just for GSON in `SharedPreferencesProjectsRepository` that has a non null ID. That would let `Project` have `String?` id.

I've been thinking it would be a good idea to use approach 1 for the forms/instances repos as well so would probably lean towards that.

Reply:
The default param value here could still just be `null` though right? In `getSettingsId` you'd just do `if (projectId == null...`.

In terms of running into problems with `Project#uuid` being null it sounds like a sign that we probably shouldn't use the same class as the "domain" object as we do to store the project. I think there are two options which let us avoid a constant:

1. Use sub-types (or if you're fancy "algebraic data types") to represent the project before and after it's saved:

    ```kotlin
    interface Project {

        val name: String

        data class New(override val name: String) : Project
        data class Existing(val uuid: String, override val name: String) : Project
    }
    ```

    In this world `ProjectsRepository#save` could take a project (as we usually do insert and update from save in our other repos) and return `Projects.Existing`.

2. Use a private class just for GSON in `SharedPreferencesProjectsRepository` that has a non-null uuid.

While I think 2 is probably always a good idea (as it ensures your storage mechanism doesn't affect your domain object) it might be overkill here. I had been thinking we'd probably want to use the kind of approach laid out in 1 for forms and instances as we have a similar problem there with IDs being assigned at save time. What do you think?

Reply:
What you said makes sense but I'm not totally convinced because:
- if we decide to migrate default preferences and always use prefixes we won't need that null and then both objects would be equal so I think it wouldn't make sense to keep both of them
- even with the current implementation the role of Project.New would be small since we build it just to pass to add() method in ProjectsRepository, we could either way pass just values and need only Project.Existing

given that I would vote for postponing this decision and make it after we decide what to do with default preferences. What do you think?

Reply:
> given that I would vote for postponing this decision and make it after we decide what to do with default preferences. What do you think?

Yeah fair. Let's get there and we can revise this.